platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java

```diff
@@ -1089,24 +1098,56 @@ public void removeAllMarkers() {
      */
     public void setMapChangeListener(@Nullable final MapChangeListener listener) {
         mapChangeListener = listener;
+        mapRegionChangeStateManager = new MapRegionChangeStateManager(mapChangeListener);
     }
 
-    //Convenience member functions
-    private void onRegionWillChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionWillChange(animated);
+    private static class MapRegionChangeStateManager {
+
+        enum MapRegionChangeState {
+            IDLE,
+            JUMPING,
+            ANIMATING,
         }
-    }
 
-    private void onRegionDidChange(boolean animated) {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionDidChange(animated);
+        WeakReference<MapChangeListener> mapChangeListenerWeakReference;
+
+        MapRegionChangeStateManager(@Nullable MapChangeListener listener) {
+            mapChangeListenerWeakReference = new WeakReference<>(listener);
         }
-    }
 
-    private void onRegionIsChanging() {
-        if (mapChangeListener != null) {
-            mapChangeListener.onRegionIsChanging();
+        MapRegionChangeState currentState = MapRegionChangeState.IDLE;
+
+        void setMapRegionState(MapRegionChangeState state) {
+
+            MapChangeListener mapChangeListener = mapChangeListenerWeakReference.get();
+            if (mapChangeListener != null) {
+                switch (currentState) {
+                    case IDLE:
+                        if (state == MapRegionChangeState.JUMPING) {
+                            mapChangeListener.onRegionWillChange(false);
+                        } else if (state == MapRegionChangeState.ANIMATING){
+                            mapChangeListener.onRegionWillChange(true);
+                        }
+                        break;
+                    case JUMPING:
+                        if (state == MapRegionChangeState.IDLE) {
+                            mapChangeListener.onRegionDidChange(false);
+                        } else if (state == MapRegionChangeState.ANIMATING) { // e.g. panning to fling
```

Comment:
I'm not sure that `JUMPING` -> `ANIMATING` is a valid transition. This would mean we produce "is changing" calls with no preceding "will change animated", which seems wrong. Also, panning currently sets the state to `ANIMATING` (though maybe it _should_ be `JUMPING`).

Reply:
I feel these gesture responders should be `ANIMATING`.

Reply:
`JUMPING` state should imply any change which results in sudden change in map view:
eg. a cameraUpdate call with duration set to 0!